testonly/hammer/mapreplay/main.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// mapreplay replays a log of Trillian Map requests.
+package main
+
+import (
+	"context"
+	"flag"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian"
+	"github.com/google/trillian/testonly/hammer"
+	"google.golang.org/grpc"
+)
+
+var (
+	mapIDs     = flag.String("map_ids", "", "Comma-separated list of mapID:mapID pairs to convert")
+	rpcServer  = flag.String("rpc_server", "", "Server address:port; leave blank for dry-run mode")
+	replayFrom = flag.String("replay_from", "", "File to record operations in")
+)
+
+func main() {
+	flag.Parse()
+	ctx := context.Background()
+
+	if *replayFrom == "" {
+		glog.Exit("Need --replay_from option")
+	}
+	f, err := os.Open(*replayFrom)
+	if err != nil {
+		glog.Exitf("Failed to open replay file: %v", err)
+	}
+
+	var cl trillian.TrillianMapClient
+	if *rpcServer != "" {
+		c, err := grpc.Dial(*rpcServer, grpc.WithInsecure())
+		if err != nil {
+			glog.Exitf("Failed to create map client conn: %v", err)
+		}
+		cl = trillian.NewTrillianMapClient(c)
+	}
+
+	pairRE := regexp.MustCompile(`(\d+):(\d+)`)
```

Comment:
Do you think this bit of manual parsing could be replaced with a proto decode of a map?

Reply:
Not really -- the input is a command-line option and so text proto input would be a) too verbose and b) too multi-line.

Reply:
(On a future pass it would be good to make the maphammer do its own CreateTree, capture that in the journal, and do the treeID mapping automatically on replay of something with CreateTree req/rsp pairs in it.  But that's for another day.)

Reply:
ah, ok.